cunumeric/array.py

```diff
@@ -3728,3 +3754,70 @@ def _perform_where(cls, mask, one, two) -> ndarray:
         out = ndarray(shape=out_shape, dtype=common_type, inputs=args)
         out._thunk.where(mask, one, two)
         return out
+
+    @classmethod
+    def _perform_scan(
+        cls, op, src, axis=None, dtype=None, out=None, nan0=False
+    ):
+        if dtype is None:
+            if out is None:
+                if np.issubdtype(src.dtype, np.integer):
+                    # Set dtype to default platform integer
+                    dtype = np.int_
+                else:
+                    dtype = src.dtype
+            else:
+                dtype = out.dtype
+        if (
+            np.issubdtype(src.dtype, np.floating) or np.iscomplexobj(src)
+        ) and np.issubdtype(dtype, np.integer):
+            # Needs changes to convert()
+            raise NotImplementedError(
+                "Integer output types currently not supported for "
+                "floating/complex inputs"
+            )
+        if np.issubdtype(dtype, np.complexfloating) and nan0 is True:
+            # Currently not behaving correctly
+            raise NotImplementedError(
+                "nancumsum and nancumprod currently do not "
+                "supported complex types"
+            )
+        if out is not None:
+            if dtype != out.dtype:
+                # if out array is specified, its type overrules dtype
+                dtype = out.dtype
+            if axis is not None:
+                assert out.shape == src.shape
+            else:
+                assert out.ndim == 1
+                assert out.size == src.size
+            if dtype == src.dtype:
+                out = convert_to_cunumeric_ndarray(out)
```

Comment:
out is not mandatory, so it can be None, so I assumed I can't use boilerplate and should convert it manually (similar to how choose() is done). Is that not correct?

Reply:
I believe `add_boilerplate` handles optional `out`

https://github.com/nv-legate/cunumeric/blob/8e60ec075269f1b70f74cdafcec801494d68ca23/cunumeric/array.py#L55-L56

Looking at the code this seems to be the case, `out` is only handled if it is actually present in `kwargs`

https://github.com/nv-legate/cunumeric/blob/8e60ec075269f1b70f74cdafcec801494d68ca23/cunumeric/array.py#L95-L103